File: udp/udp/udp_server.py
import socket
import time

class UDPServer:
    def __init__(self, ip="0.0.0.0", port=8080, buffer_size=1024, total_packets=200):
        self.udp_ip = ip
        self.udp_port = port
        self.buffer_size = buffer_size
        self.total_packets = total_packets
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.sock.bind((self.udp_ip, self.udp_port))
        print(f"Servidor UDP ouvindo na porta {self.udp_port}")

    def listen(self):
        while True:
            data, addr = self.sock.recvfrom(self.buffer_size)
            print(f"Mensagem recebida de {addr}: {data.decode()}")
            self.send_total_packets(addr)
            self.send_packets(addr)

            while True:
                missing_request, addr = self.sock.recvfrom(self.buffer_size)
                missing_packets = missing_request.decode().split(',')
                if missing_packets[0] == 'sair':
                    print(f"Conexão com {addr} encerrada.")
                    break
                self.send_missing_packets(missing_packets, addr)

    def send_total_packets(self, addr):
        self.sock.sendto(str(self.total_packets).encode(), addr)

    def send_packets(self, addr):
        for i in range(self.total_packets):
            response = f"Pacote {i+1}"
            self.sock.sendto(response.encode(), addr)
            time.sleep(0.1)
        print(f"{self.total_packets} pacotes enviados.")

    def send_missing_packets(self, missing_packets, addr):
        for packet in missing_packets:
            response = f"Pacote {packet}"
            self.sock.sendto(response.encode(), addr)
        print(f"Pacotes solicitados enviados: {missing_packets}")

if __name__ == "__main__":
    server = UDPServer()
    server.listen()

File: udp/udp/udp_server.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <arpa/inet.h>
#include <unistd.h>

#define SERVER_PORT 8080
#define BUFFER_SIZE 1024
#define TOTAL_PACKETS 200

void send_total_packets(int sockfd, struct sockaddr_in client_addr) {
    char buffer[BUFFER_SIZE];
    sprintf(buffer, "%d", TOTAL_PACKETS);
    sendto(sockfd, buffer, strlen(buffer), 0, (struct sockaddr*)&client_addr, sizeof(client_addr));
}

void send_packets(int sockfd, struct sockaddr_in client_addr) {
    char buffer[BUFFER_SIZE];
    for (int i = 0; i < TOTAL_PACKETS; i++) {
        sprintf(buffer, "Pacote %d", i + 1);
        sendto(sockfd, buffer, strlen(buffer), 0, (struct sockaddr*)&client_addr, sizeof(client_addr));
        usleep(100000);
    }
    printf("%d pacotes enviados.\n", TOTAL_PACKETS);
}

void send_missing_packets(int sockfd, struct sockaddr_in client_addr, char* missing_packets) {
    char* token = strtok(missing_packets, ",");
    while (token != NULL) {
        char buffer[BUFFER_SIZE];
        sprintf(buffer, "Pacote %s", token);
        sendto(sockfd, buffer, strlen(buffer), 0, (struct sockaddr*)&client_addr, sizeof(client_addr));
        token = strtok(NULL, ",");
    }
    printf("Pacotes solicitados enviados: %s\n", missing_packets);
}

int main() {
    int sockfd;
    struct sockaddr_in server_addr, client_addr;
    char buffer[BUFFER_SIZE];
    socklen_t addr_len = sizeof(client_addr);

    sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        perror("Erro ao criar socket");
        return 1;
    }

    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(SERVER_PORT);
    server_addr.sin_addr.s_addr = INADDR_ANY;

    if (bind(sockfd, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        perror("Erro ao fazer bind");
        close(sockfd);
        return 1;
    }

    printf("Servidor UDP ouvindo na porta %d\n", SERVER_PORT);

    while (1) {
        recvfrom(sockfd, buffer, BUFFER_SIZE, 0, (struct sockaddr*)&client_addr, &addr_len);
        printf("Mensagem recebida de %s:%d: %s\n", inet_ntoa(client_addr.sin_addr), ntohs(client_addr.sin_port), buffer);
        send_total_packets(sockfd, client_addr);
        send_packets(sockfd, client_addr);

        while (1) {
            recvfrom(sockfd, buffer, BUFFER_SIZE, 0, (struct sockaddr*)&client_addr, &addr_len);
            buffer[strcspn(buffer, "\n")] = '\0';
            if (strcmp(buffer, "sair") == 0) {
                printf("Conexão com %s:%d encerrada.\n", inet_ntoa(client_addr.sin_addr), ntohs(client_addr.sin_port));
                break;
            }
            send_missing_packets(sockfd, client_addr, buffer);
        }
    }

    close(sockfd);
    return 0;
}